### PR TITLE
feat: add base256emoji

### PR DESCRIPTION
**Pull request by Jorropo**
I think this should be merged because :
- It's fun and allows us have an almost endless of fun in docs and trolling opportunities.
- Most of the NFT community likes useless differentiation features, why not giving what they want ?
- It's a good test for unicode based encodings in current implementations.
- It makes for a great education tool about what the spirit of multiformats is about (I found that some new people over sacralise what a CID is, if we allows dumb stuff like that I think it will be helpfull to explain to people that things are really extensible and you can make whatever you want on top of it).
- It's not taking valuable prefix space (the 🚀 preffix is very unlikely to ever be used by anything else).
- Implementations that doesn't support unicode will just error with codec not found (since the first byte of 🚀 is not valid ascii nor unicode it would be wrong from us to use as a prefix).

### TL;DR:
I think it has a very slight positive impact and no negative impact, let's merge this. :)

### CODE REVIEW

**Comment by Stebalien** [comment]
Hm. My first thought is that the readme needs some pithy examples.

Other than that :heavy_check_mark:.

**Comment by ribasushi** [comment]
I love this! A note: you use 🚀 both as "multibase prefix" and as "digit 0", perhaps disambiguating the two is worthwhile 🤷

**Comment by Jorropo** [comment]
@ribasushi 
> you use rocket both as "multibase prefix" and as "digit 0", perhaps disambiguating the two is worthwhile

I don't really get why that an issue, whats the rational behind this ?
I believe the only thing we was trying to do is:
- we should attempt to use prefixes that doesn't encode in smaller / alternative bases (which I'm certain to have covered)
- the prefix should be in the alphabet (which again is)

Would swapping 🚀 (0) and 🪐 (5) in the table by enough of a fix ?

File: rfcs/Base256Emoji.md [hunk 1]
```diff
@@ -0,0 +1,543 @@
...
+-+-+-
+🚀|U+1F680|0
+-+-+-
+🙬|U+1F66C|1
+-+-+-
+🙭|U+1F66D|2
+-+-+-
+🙮|U+1F66E|3
+-+-+-
+🙯|U+1F66F|4
```
Comment (Stebalien) [inline rfcs/Base256Emoji.md:34]: not supported

File: rfcs/Base256Emoji.md [hunk 2]
```diff
@@ -0,0 +1,543 @@
...
+-+-+-
+🌍|U+1F30D|17
+-+-+-
+🌏|U+1F30F|18
+-+-+-
+🌎|U+1F30E|19
+-+-+-
+☉|U+2609|20
+-+-+-
+☼|U+263C|21
```
Comment (Stebalien) [inline rfcs/Base256Emoji.md:68]: Not supported

File: rfcs/Base256Emoji.md [hunk 3]
```diff
@@ -0,0 +1,543 @@
...
+-+-+-
+☀|U+2600|22
+-+-+-
+💻|U+1F4BB|23
+-+-+-
+🖥|U+1F5A5|24
+-+-+-
+🖳|U+1F5B3|25
```
Comment (Stebalien) [inline rfcs/Base256Emoji.md:76]: Not supported.

File: rfcs/Base256Emoji.md [hunk 4]
```diff
@@ -0,0 +1,543 @@
...
+-+-+-
+👍|U+1F44D|37
+-+-+-
+😅|U+1F605|38
+-+-+-
+👏|U+1F44F|39
+-+-+-
+😁|U+1F601|40
+-+-+-
+♥|U+2665|41
```
Comment (Stebalien) [inline rfcs/Base256Emoji.md:108]: Easily confused with heart emoji?

File: rfcs/Base256Emoji.md [hunk 5]
```diff
@@ -0,0 +1,543 @@
...
+-+-+-
+😇|U+1F607|59
+-+-+-
+🌹|U+1F339|60
+-+-+-
+🤦|U+1F926|61
+-+-+-
+🎉|U+1F389|62
+-+-+-
+‼|U+203C|63
```
Comment (Stebalien) [inline rfcs/Base256Emoji.md:152]: Not a real emoji.

**Comment by Jorropo** [comment]
I have two counter arguments :
- I like rockets, the more rockets there is the happier I am. (It's not like it's taking the place of an other valuable emoji that could be represented, see the go impl code comments for emoji rational (btw if anyone has ideas for important emojies to include send it pls))
- Other bases don't do that, most of them use their biggest glyph as prefix. (Here I'm doing the reverse so inlined blocks have at least two rockets (key + identity multihash) because cid inlining is very cool and because all of the collision with smaller base thing is not an issue this time).

File: rfcs/Base256Emoji.md [hunk 6]
```diff
@@ -0,0 +1,543 @@
...
+-+-+-
+🤢|U+1F922|247
+-+-+-
+👅|U+1F445|248
+-+-+-
+💡|U+1F4A1|249
+-+-+-
+💩|U+1F4A9|250
+-+-+-
+⁉|U+2049|251
```
Comment (Stebalien) [inline rfcs/Base256Emoji.md:528]: Not really an emoji.

**Review by Stebalien — COMMENTED** [review]
General emoji choice comments. I'd also consider trying to make emoji's as unique as possible.

(I mean, if we're going to paint the bikeshed, we might as well paint it well)

**Comment by Jorropo** [comment]
> (I mean, if we're going to paint the bikeshed, we might as well paint it well)

The only real usecase I see is NFT artists bruteforcing cool emoji patern for their collection.

So I belive we need rockets and moons.
Then for the rest I've figurated that nft artists likely like the same emojies as the average peoples so I took a top list from unicode's popularity count.

**Comment by vmx** [comment]
I'm also in favour of using an emoji as prefix which is in the set of the 256 emojis

> because cid inlining is very cool

It is not and I hope it's used less and less. It causes more trouble then good. The only use case I see is for data where the hash is bigger than the actual data. Though even then you're probably better off with other solutions, e.g. combining some of your data into bigger chunks.

**Comment by BigLep** [comment]
@Jorropo : are you able to incorporate the input here, particularly around not supported emojis?

**Comment by Jorropo** [comment]
@BigLep 
> are you able to incorporate the input here, particularly around not supported emojis?

Yes, at first I didn't really wanted to, but since it seems that everyone think emojies that work is an important thing yes.
It's just really unimportant and havn't attempted to give me time to work on this yet.

File: rfcs/Base256Emoji.md (hunk 2 above)
Comment (Jorropo) [reply rfcs/Base256Emoji.md:68]: removed

File: rfcs/Base256Emoji.md (hunk 3 above)
Comment (Jorropo) [reply rfcs/Base256Emoji.md:76]: removed

File: rfcs/Base256Emoji.md (hunk 5 above)
Comment (Jorropo) [reply rfcs/Base256Emoji.md:152]: removed

File: rfcs/Base256Emoji.md (hunk 6 above)
Comment (Jorropo) [reply rfcs/Base256Emoji.md:528]: removed

**Comment by Jorropo** [comment]
New fixed version pushed.

Pls could you merge with `git merge` (not `rebase`) so the commit hash wont change and `go-multibase`'s submodule wont need yet an other push.

File: multibase.csv [hunk 7]
```diff
@@ -1,25 +1,26 @@
...
+base36,            k,    base36 [0-9a-z] case-insensitive - no padding,                draft
+base36upper,       K,    base36 [0-9a-z] case-insensitive - no padding,                draft
+base58btc,         z,    base58 bitcoin,                                               default
+base58flickr,      Z,    base58 flicker,                                               candidate
+base64,            m,    rfc4648 no padding,                                           default
+base64pad,         M,    rfc4648 with padding - MIME encoding,                         candidate
+base64url,         u,    rfc4648 no padding,                                           default
+base64urlpad,      U,    rfc4648 with padding,                                         default
+proquint,          p,    PRO-QUINT https://arxiv.org/html/0901.4016,                   draft
+base256emoji,      🚀,   base256 with custom alphabet using variable-sized-codepoints, draft
```
Comment (rvagg) [inline multibase.csv:26]: the most triggering thing about this is the fact that the emoji breaks the fixed-width table:
Screenshot 2022-06-08 at 3 18 56 pm

**Review by rvagg — APPROVED** [review]
cool, got an implementation for this yet?

**Comment by Jorropo** [comment]
yes https://github.com/multiformats/go-multibase/pull/46
I'm just cleaning it up and also applying the requested changes

File: multibase.csv (hunk 7 above)
Comment (Jorropo) [reply multibase.csv:26]: I know sorry (no I love that).

Ju*ST G*eT *A PR*Oper m*onO* C*aSE* FO*Nt.*

Comment (Jorropo) [reply multibase.csv:26]: actually I'm one space to short

Comment (Jorropo) [reply multibase.csv]: Now actually monospace correct (if you had a monospace emoji it wouldn't break).

**Comment by Jorropo** [comment]
https://github.com/multiformats/go-multibase/pull/46 is now ready for review too.

On my side it's RFM

**Comment by rvagg** [comment]
Merge commits really grind my gears .. but since you insist and there's recent precedent in this repo I'll hold my nose

https://github.com/multiformats/js-multiformats/pull/187 for JS impl
